File: LeetCode/1799. Maximize Score After N Operations/solution.py
class Solution:
    def maxScore(self, nums: List[int]) -> int:
        memo = {}
        n = len(nums)//2
        def gcd(a, b):
            a, b = max(a,b), min(a,b)
            if b == 0:
                return a
            return gcd(b, a%b)

        def dfs(cur,idx):
            # print(cur,idx)
            if cur in memo:
                return memo[cur]
            if idx == n:
                return 0
            res = 0
            # print(cur,idx,res)
            for i in range(2*n):
                if cur & (1<<i):
                    for j in range(i+1,2*n):
                        if cur & (1<<j):
                            tmp = gcd(nums[i],nums[j])
                            # print(i,j,nums[i],nums[j],tmp)
                            res = max(res, (idx+1)*tmp+dfs(cur^(1<<i)^(1<<j), idx+1))
            # print(cur,idx,res)
            memo[cur] = res
            return res 

        ans = dfs((1<<(2*n))-1,0)
        # print(memo)
        return ans

File: LeetCode/1799. Maximize Score After N Operations/solution.cpp
// class Solution
// {
// public:
//     int maxScore(vector<int> &nums)
//     {
//         int n = nums.size();
//         vector<int> dp(1 << n, 0);
//         for (int i = 1; i < (1 << n); i++)
//         {
//             int m = __builtin_popcount(i);
//             if (m & 1)
//                 continue;
//             for (int j = 0; j < n; j++)
//             {
//                 if ((i >> j) & 1){
//                     for (int k = j + 1; k < n; k++)
//                     {
//                         if ((i >> k) & 1)
//                         {
//                             // cout << i << " " << i - (1 << j) - (1 << k) << " " << j << " " << k << endl;
//                             dp[i] = max(dp[i], dp[i - (1 << j) - (1 << k)] + __gcd(nums[j], nums[k]) * m / 2);
//                         }
//                     }
//                 }
                
//             }
//         }
//         return dp.back();
//     }
// };


class Solution
{
public:
    int maxScore(vector<int> &nums)
    {
        int n = nums.size();
        vector<int> dp(1 << n, 0);
        for (int i = 1; i < (1 << n); i++)
        {
            int m = __builtin_popcount(i);
            if (m & 1)
                continue;
            for (int j = 0; j < n; j++)
            {
                if (((i >> j) & 1) == 0)
                    continue;
                for (int k = j + 1; k < n; k++)
                {
                    if (((i >> k) & 1) == 0)
                        continue;
                    // cout << i << " " << i - (1 << j) - (1 << k) << " " << j << " " << k << endl;
                    dp[i] = max(dp[i], dp[i - (1 << j) - (1 << k)] + __gcd(nums[j], nums[k]) * m / 2);
                }
            }
        }
        return dp.back();
    }
};